src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -472,34 +470,40 @@ ucc_status_t ucc_tl_ucp_ctx_remote_populate(ucc_tl_ucp_context_t * ctx,
         uint64_t *key_sizes = PTR_OFFSET(base, sizeof(uint64_t) * nsegs * 2);
         void *    key       = PTR_OFFSET(base, sizeof(uint64_t) * nsegs * 3);
         size_t    offset    = 0;
-        int       j;
 
         if (i != rank) {
             remote_info[i] = (ucc_tl_ucp_remote_info_t *)ucc_calloc(
                 nsegs, sizeof(ucc_tl_ucp_remote_info_t), "ucp_remote_info");
             if (NULL == remote_info[i]) {
-
                 tl_error(ctx->super.super.lib, "failed to allocated %zu bytes",
                          sizeof(size_t) * size);
                 ucc_status = UCC_ERR_NO_MEMORY;
 
-                for (j = i - 1; j >= 0; j--) {
-                    if (j != rank) {
-                        ucc_free(remote_info[j]);
-                    }
-                }
-                goto fail_pack_exchange;
+                goto fail_unpack;
             }
         }
 
-        for (j = 0; j < nsegs; j++) {
-            remote_info[i][j].va_base = (void *)rvas[j];
-            remote_info[i][j].len     = lens[j];
-            remote_info[i][j].packed_key =
-                ucc_malloc(key_sizes[j], "ucp_packed_key");
-            memcpy(remote_info[i][j].packed_key, PTR_OFFSET(key, offset),
-                   key_sizes[j]);
-            offset += key_sizes[j];
+        for (seg = 0; seg < nsegs; seg++) {
+            remote_info[i][seg].va_base = (void *)rvas[seg];
+            remote_info[i][seg].len     = lens[seg];
+            remote_info[i][seg].packed_key =
+                ucc_malloc(key_sizes[seg], "ucp_packed_key");
+            if (NULL == remote_info[i][seg].packed_key) {
+                tl_error(ctx->super.super.lib, "failed to allocated %zu bytes",
+                         key_sizes[seg]);
+                ucc_status = UCC_ERR_NO_MEMORY;
+
+                for (seg = seg - 1; seg >= 0; seg--) {
```

Comment:
can't we just go to "fail_unpack" and check if remote_info[rank][i].packed_key != NULL there and release. To avoid some cleanup code duplication